libclamav/readdb.c

```diff
@@ -1524,30 +1541,56 @@ static cl_error_t cli_loadidb(FILE *fs, struct cl_engine *engine, unsigned int *
         if (matcher->group_counts[0] > 256 || matcher->group_counts[1] > 256) {
             cli_errmsg("cli_loadidb: too many icon groups!\n");
             ret = CL_EMALFDB;
-            break;
+            goto done;
         }
 
         sigs++;
     }
-    if (engine->ignored)
-        free(buffer_cpy);
 
     if (!line) {
-        cli_errmsg("cli_loadidb: Empty database file\n");
-        ret = CL_EMALFDB;
+        cli_dbgmsg("cli_loadidb: Empty database file\n");
+    } else {
+        cli_dbgmsg("cli_loadidb: %u signatures loaded\n", sigs);
+        if (signo) {
+            *signo += sigs;
+        }
     }
 
-    if (ret) {
-        cli_errmsg("cli_loadidb: Problem parsing database at line %u\n", line);
-        MPOOL_FREE(engine->mempool, matcher);
-        return ret;
+    ret = CL_SUCCESS;
+
+done:
+
+    if (NULL != buffer_cpy) {
+        free(buffer_cpy);
     }
 
-    if (signo)
-        *signo += sigs;
+    if (CL_SUCCESS != ret) {
+        for (i = 0; i < 3; i++) {
+            if (matcher->icons[i]) {
+                uint32_t j;
 
-    engine->iconcheck = matcher;
-    return CL_SUCCESS;
+                for (j = 0; j < matcher->icon_counts[i]; j++) {
+                    struct icomtr *metric = matcher->icons[i];
+                    MPOOL_FREE(engine->mempool, metric[j].name);
+                }
+                MPOOL_FREE(engine->mempool, matcher->icons[i]);
+            }
+        }
+        if (matcher->group_names[0]) {
+            for (i = 0; i < matcher->group_counts[0]; i++)
+                MPOOL_FREE(engine->mempool, matcher->group_names[0][i]);
+            MPOOL_FREE(engine->mempool, matcher->group_names[0]);
+        }
+        if (matcher->group_names[1]) {
+            for (i = 0; i < matcher->group_counts[1]; i++)
+                MPOOL_FREE(engine->mempool, matcher->group_names[1][i]);
+            MPOOL_FREE(engine->mempool, matcher->group_names[1]);
+        }
```

Comment:
If you need to make a minor change here, you're going to have to do it twice.

I'd write this as:

for (i = 0; i < 2; i++) {
    if (matcher->group_names[i]) {
        for (j = 0; j < matcher->group_counts[i]; j++)
            MPOOL_FREE(engine->mempool, matcher->group_names[i][j]);
        MPOOL_FREE(engine->mempool, matcher->group_names[i]);
    }
}